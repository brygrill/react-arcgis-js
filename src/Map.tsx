import { loadCss, loadModules } from 'esri-loader';
import * as React from 'react';

export interface IMapProps {
  itemId: string;
}

export class Map extends React.Component<IMapProps, {}> {
  public static defaultProps: Partial<IMapProps> = {
    itemId: 'f2e9b762544945f390ca4ac3671cfa72',
  };
  public componentDidMount() {
    loadModules(['esri/views/MapView', 'esri/WebMap'])
      .then(([MapView, WebMap]) => {
        const map = new WebMap({
          portalItem: {
            id: this.props.itemId,
          },
        });
        const view = new MapView({
          map,
          container: 'my-map',
        });
      })
      .catch((e: Error) => {
        console.log(e);
      });
  }
  public render() {
    return <div id="my-map" style={{ height: '100%', width: '100%' }} />;
  }
}
