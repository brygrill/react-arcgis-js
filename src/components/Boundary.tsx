import * as React from 'react';

export interface IBoundaryProps {
  onErrorContent?: any; // string or component to render on error
  onError?(error: any): void; // function that will get called on error and return the error as { error: ..., info: ...}
}

export interface IBoundaryState {
  error: boolean;
}

export class Boundary extends React.Component<IBoundaryProps, IBoundaryState> {
  public static defaultProps: Partial<IBoundaryProps> = {
    onErrorContent: 'Error loading map...',
    onError: () => {},
  };

  constructor(props: IBoundaryProps) {
    super(props);
    this.state = {
      error: false,
    };
  }

  componentDidCatch(error: Error, info: React.ErrorInfo) {
    const { onError } = this.props;
    onError!({ error, info });
    this.setState({ error: true });
  }

  render() {
    if (this.state.error) {
      return <div>{this.props.onErrorContent}</div>;
    }
    return <div>{this.props.children}</div>;
  }
}
