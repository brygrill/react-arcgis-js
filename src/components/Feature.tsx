import * as React from 'react';

import { loadFeatureLayerModule } from '../helpers';

export interface IMapPropInterface {
  add: Function;
}

export interface IFeatureProps {
  map: IMapPropInterface;
  view: object;
  url: string;
  itemId: string;
  geojson: object;
}

export interface IFeatureState {
  loading: boolean;
  error: boolean;
  featureProperties: object;
  feature: object;
}

// format the FL options based on the prop passed in
const setfeatureProperties = (props: IFeatureProps) => {
  switch (true) {
    case !!props.geojson:
      return {};
    case !!props.itemId:
      return {
        portalItem: {
          id: props.itemId,
        },
      };
    case !!props.url:
      return {
        url: props.url,
      };
    default:
      return {};
  }
};

export class Feature extends React.Component<IFeatureProps, IFeatureState> {
  constructor(props: IFeatureProps) {
    super(props);
    this.createFeature = this.createFeature.bind(this);
    this.state = {
      loading: true,
      error: false,
      featureProperties: setfeatureProperties(props),
      feature: {},
    };
  }

  async createFeature() {
    try {
      const FeatureLayer = await loadFeatureLayerModule();
      const feature = new FeatureLayer(this.state.featureProperties);
      feature.load().then((feature: object) => {
        this.props.map.add(feature);
        this.setState({ loading: false, feature });
      });
    } catch (error) {
      this.setState({ loading: false, error: true });
    }
  }

  componentDidMount() {
    this.createFeature();
  }

  render() {
    const childrenWithProps = React.Children.map(this.props.children, child => {
      const childEl = child as React.ReactElement<any>;
      return React.cloneElement(childEl, {
        loading: this.state.loading,
        error: this.state.error,
        feature: this.state.feature,
      });
    });
    return <div>{childrenWithProps}</div>;
  }
}
