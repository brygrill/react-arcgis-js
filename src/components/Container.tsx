import * as React from 'react';

import { Boundary } from './Boundary';

export interface IContainerProps {
  height: string;
  width: string;
  hidden?: boolean;
  id: string;
  onErrorContent: any;
  children?: any;
}

export const Container = (props: IContainerProps) => {
  const { height, width, hidden, children} = props;
  return (
    <Boundary onErrorContent={props.onErrorContent}>
      <div style={{ height, width }}>
        <div
          id={props.id}
          style={{ height: '100%', width: '100%', display: hidden ? 'none' : null }}
        />
        {children}
      </div>
    </Boundary>
  );
};
