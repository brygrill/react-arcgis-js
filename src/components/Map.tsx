import { loadModules } from 'esri-loader';
import * as React from 'react';

import { Container } from './Container';

import { BaseMapOptions } from '../helpers';

export interface IMapProps {
  onLoading?: any; // string or component to render while loading map
  onError?: any; //  string or component to render on error
  itemId?: string; // https://developers.arcgis.com/javascript/latest/api-reference/esri-WebMap.html#portalItem
  baseMap?: BaseMapOptions; // https://developers.arcgis.com/javascript/latest/api-reference/esri-Map.html#basemap
  height: string; // height of map container in px or %
  width: string; // width of map container in px or %
  center?: Array<Number>; // https://developers.arcgis.com/javascript/latest/api-reference/esri-views-MapView.html#center
  zoom?: number; // https://developers.arcgis.com/javascript/latest/api-reference/esri-views-MapView.html#zoom
  webGL?: boolean; // will load dojo config to enable WebGL Feature Layers
}

export interface IMapState {
  loading: boolean;
  error: boolean;
  containerId: string;
  modules: Array<string>;
  mapOptions: object;
  map: object;
  view: object;
  moduleOptions: object;
}

export class Map extends React.Component<IMapProps, IMapState> {
  public static defaultProps: Partial<IMapProps> = {
    onLoading: 'Loading...',
    onError: 'Error loading map...',
    baseMap: BaseMapOptions.streets_navigation_vector,
    height: '500px',
    width: '100%',
    center: [-122.41, 37.77],
    zoom: 10,
  };

  constructor(props: IMapProps) {
    super(props);
    this.state = {
      loading: true,
      error: false,
      containerId: Math.random()
        .toString(36)
        .substr(2, 15),
      modules: props.itemId
        ? ['esri/views/MapView', 'esri/WebMap']
        : ['esri/views/MapView', 'esri/Map'],
      moduleOptions: props.webGL
        ? {
            dojoConfig: {
              has: {
                'esri-featurelayer-webgl': 1,
              },
            },
          }
        : {},
      mapOptions: props.itemId
        ? {
            portalItem: {
              id: this.props.itemId,
            },
          }
        : {
            basemap: this.props.baseMap,
          },
      map: {},
      view: {},
    };
  }

  createMap = () => {
    console.log('create map');
    loadModules(this.state.modules, this.state.moduleOptions)
      .then(([MapView, Map]) => {
        const map = new Map(this.state.mapOptions);
        const view = new MapView({
          map,
          container: this.state.containerId,
          center: this.props.center,
          zoom: this.props.zoom,
        });
        console.log(map);
        this.setState({ loading: false, map, view });
      })
      .catch(() => {
        this.setState({ loading: false, error: true });
      });
  };

  componentDidMount() {
    console.log('map cDM');
    this.createMap();
  }

  render() {
    // on loading render loader
    if (this.state.loading) {
      return (
        <Container
          height={this.props.height}
          width={this.props.width}
          onError={this.props.onError}
          id={this.state.containerId}
          hidden={this.state.loading}
        >
          {this.props.onLoading}
        </Container>
      );
    }

    // if error render error msg
    if (this.state.error) {
      return (
        <Container
          height={this.props.height}
          width={this.props.width}
          onError={this.props.onError}
          id={this.state.containerId}
          hidden={this.state.error}
        >
          {this.props.onError}
        </Container>
      );
    }

    // when map is ready, pass props to children
    const childrenWithProps = React.Children.map(this.props.children, child => {
      const childEl = child as React.ReactElement<any>;
      return React.cloneElement(childEl, { ...this.state });
    });

    // render the map and children
    return (
      <Container
        height={this.props.height}
        width={this.props.width}
        onError={this.props.onError}
        id={this.state.containerId}
      >
        {childrenWithProps}
      </Container>
    );
  }
}
