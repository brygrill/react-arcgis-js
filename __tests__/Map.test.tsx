import { configure, mount, shallow } from 'enzyme';
import * as Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import * as React from 'react';
import * as ReactDOM from 'react-dom';
import * as renderer from 'react-test-renderer';

import { Map } from '../src/components/Map';

configure({ adapter: new Adapter() });

describe('<Map />', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
      <Map height="500px" width="100%" onError="Error!">
        <div>Child Component</div>
      </Map>,
      div,
    );
  });

  it('should initially display loading msg', () => {
    const tree = renderer
      .create(
        <Map height="500px" width="100%">
          <div>Child Component</div>
        </Map>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should load map on mount', async () => {
    const component = await mount(
      <Map height="500px" width="100%">
        <div>Child Component</div>
      </Map>,
    );
    // const component = shallow(
    //   <Map height="500px" width="100%">
    //     <div>Child Component</div>
    //   </Map>,
    // );
    // await component.instance().componentDidMount();
    component.update()
    // console.log(component.state('map'));
    // expect(component.state('map')).toBeTruthy();
    // expect(toJson(component)).toMatchSnapshot();
    console.log(component.state());
  });
});
