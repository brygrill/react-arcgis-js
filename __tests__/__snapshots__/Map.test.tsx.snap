// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Map /> should initially display loading msg 1`] = `
<div>
  <div
    style={
      Object {
        "height": "500px",
        "width": "100%",
      }
    }
  >
    <div
      id="ytjc02stj4e"
      style={
        Object {
          "display": "none",
          "height": "100%",
          "width": "100%",
        }
      }
    />
    Loading...
  </div>
</div>
`;
